### PR TITLE
FEAT: support pygloo on Windows

### LINKED ISSUE #33: ENH: collective communication supports windows
Note that the issue tracker is NOT the place for general support.  For
discussions about development, questions about usage, or any general questions,
contact us on https://discuss.xorbits.io/.

Related #22

### PR DESCRIPTION
**Pull request by YibinLiu666**
## Related issue number

Fixes #33 

## Check code requirements

- [ ] tests added / passed (if needed)
- [ ] Ensure all linting tests pass

### CODE REVIEW

File: cpp/collective/rendezvous/CMakeLists.txt [hunk 1]
```diff
@@ -5,13 +5,16 @@ project(
   VERSION 0.0.1
   LANGUAGES CXX)
 
-set(CMAKE_CXX_STANDARD 20)
+set(CMAKE_CXX_STANDARD 17)
```
Comment (ChengjieLi28) [inline cpp/collective/rendezvous/CMakeLists.txt:8]: Only for windows C++17. Other platform remain C++ 20. And add a comment to explain why windows needs C++17.

File: cpp/collective/rendezvous/CMakeLists.txt [hunk 2]
```diff
@@ -5,13 +5,16 @@ project(
...
 
 include_directories(include)
 
 add_library(
   StoreLib
   include/error.h
+  include/win32-headers.h
```
Comment (ChengjieLi28) [inline cpp/collective/rendezvous/CMakeLists.txt:15]: naming this file as ``win32_headers.h``.

File: cpp/collective/rendezvous/CMakeLists.txt [hunk 3]
```diff
@@ -5,13 +5,16 @@ project(
...
+  include/CallOnce.h
+  include/WinSockUtils.hpp
```
Comment (ChengjieLi28) [inline cpp/collective/rendezvous/CMakeLists.txt:17]: naming issue.

File: cpp/collective/rendezvous/include/utils.hpp [hunk 4]
```diff
@@ -19,18 +19,275 @@ limitations under the License. */
...
+decltype(auto) torchCheckMsgImpl(const char * /*msg*/, const Args &...args) {
+    return ::xoscar::str(args...);
+}
+inline const char *torchCheckMsgImpl(const char *msg) { return msg; }
+// If there is just 1 user-provided C-string argument, use it.
+inline const char *torchCheckMsgImpl(const char * /*msg*/, const char *args) {
+    return args;
+}
+
+#define TORCH_CHECK_MSG(cond, type, ...)                                       \
```
Comment (ChengjieLi28) [inline cpp/collective/rendezvous/include/utils.hpp:265]: Change the name please. Check all the other places.

File: python/setup.py [hunk 5]
```diff
@@ -166,6 +167,27 @@ def copy_extensions_to_source(self):
...
 
+                # Always copy, even if source is older than destination, to ensure
+                # that the right extensions for the current Python/platform are
+                # used.
+                copy_file(
+                    src_filename, dest_filename, verbose=self.verbose,
+                    dry_run=self.dry_run
+                )
+                if ext._needs_stub:
+                    self.write_stub(package_dir or os.curdir, ext, True)
```
Comment (ChengjieLi28) [inline python/setup.py:178]: Why change here?

File: python/xoscar/collective/tests/test_pygloo_tcp_store.py [hunk 6]
```diff
@@ -12,12 +12,15 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# When testing Gloo on Windows, the number of processes used for testing
+# must be less than or equal to the number of cores of the testing device.
+
```
Comment (ChengjieLi28) [inline python/xoscar/collective/tests/test_pygloo_tcp_store.py:17]: Should not add these lines here. Add these to the related tests.

File: python/xoscar/collective/tests/test_pygloo.py [hunk 7]
```diff
@@ -22,6 +22,10 @@
 
 system_name = platform.system()
 
+# When running tests on Windows, some tests may fail due to file system
+# permission issues. Therefore, we skip the relevant tests related to
+# filestore on Windows.
+
```
Comment (ChengjieLi28) [inline python/xoscar/collective/tests/test_pygloo.py:28]: You can use ``pytest.skip`` to skip some test on windows, and add comments there.

Comment (YibinLiu666) [reply python/xoscar/collective/tests/test_pygloo.py:28]: There will be some error on windows when testing `filestore` on windows, so I just run `test_pygloo.py` on `unix`

File: cpp/collective/rendezvous/include/utils.hpp [hunk 8]
```diff
@@ -19,18 +19,273 @@ limitations under the License. */
...
+void xoscarCheckFail(const char *func,
+                     const char *file,
+                     uint32_t line,
+                     const char *msg) {
+    throw ::xoscar::Error({func, file, line}, msg);
+}
+
+#if defined(__GNUC__) || defined(__ICL) || defined(__clang__)
+#    define C10_LIKELY(expr) (__builtin_expect(static_cast<bool>(expr), 1))
+#    define C10_UNLIKELY(expr) (__builtin_expect(static_cast<bool>(expr), 0))
```
Comment (codingl2k1) [inline cpp/collective/rendezvous/include/utils.hpp:214]: It's better to rename the macro prefix. `C10_UNLIKELY` to `XOSCAR_UNLIKELY`.

File: cpp/collective/rendezvous/src/socket.cpp [hunk 9]
```diff
@@ -514,10 +515,8 @@ bool SocketListenOp::tryListen(const ::addrinfo &addr) {
     // Here we follow the recommendation of Microsoft and use the non-standard
     // SO_EXCLUSIVEADDRUSE flag instead.
     if (!socket_->enableExclusiveAddressUse()) {
-        xoscar_WARNING(
```
Comment (codingl2k1) [inline cpp/collective/rendezvous/src/socket.cpp:517]: I found a `xoscar_WARNING` defined in [cpp/collective/rendezvous/include/socket.h](https://github.com/xorbitsai/xoscar/pull/51/files#diff-819cfc9e2984ed2c8a4b6266db58acfd9b7310d0dfa492d697f4a230a7d0a813). Why changing this `xoscar_WARNING` to std::cout?

File: cpp/collective/rendezvous/include/utils.hpp (hunk 8 above)
Comment (YibinLiu666) [reply cpp/collective/rendezvous/include/utils.hpp]: Done

File: cpp/collective/rendezvous/include/utils.hpp [hunk 10]
```diff
@@ -19,18 +19,279 @@ limitations under the License. */
...
+                     uint32_t line,
+                     const char *msg) {
+    throw ::xoscar::Error({func, file, line}, msg);
+}
+
+#if defined(__GNUC__) || defined(__ICL) || defined(__clang__)
+#    ifndef XOSCAR_LIKELY
+#        define XOSCAR_LIKELY(expr)                                            \
+            (__builtin_expect(static_cast<bool>(expr), 1))
+#        define XOSCAR_UNLIKELY(expr)                                          \
```
Comment (codingl2k1) [inline cpp/collective/rendezvous/include/utils.hpp:216]: Why not use just one define?

**Review by codingl2k1 — APPROVED** [review]
LGTM

**Review by ChengjieLi28 — APPROVED** [review]

**Pull request merged**